omp/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -62,22 +69,62 @@ namespace amgx_pgm {
 template <typename IndexType>
 void match_edge(std::shared_ptr<const OmpExecutor> exec,
                 const Array<IndexType> &strongest_neighbor,
-                Array<IndexType> &agg) GKO_NOT_IMPLEMENTED;
+                Array<IndexType> &agg)
+{
+    auto agg_vals = agg.get_data();
+    auto strongest_neighbor_vals = strongest_neighbor.get_const_data();
+#pragma omp parallel for
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        if (agg_vals[i] == -1) {
+            auto neighbor = strongest_neighbor_vals[i];
+            if (neighbor != -1 && strongest_neighbor_vals[neighbor] == i &&
+                i < neighbor) {
+                // Use the smaller index as agg point
+                agg_vals[i] = i;
+                agg_vals[neighbor] = i;
+            }
+        }
+    }
+}
 
 GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_AMGX_PGM_MATCH_EDGE_KERNEL);
 
 
 template <typename IndexType>
 void count_unagg(std::shared_ptr<const OmpExecutor> exec,
-                 const Array<IndexType> &agg,
-                 IndexType *num_unagg) GKO_NOT_IMPLEMENTED;
+                 const Array<IndexType> &agg, IndexType *num_unagg)
+{
+    IndexType unagg = 0;
+#pragma omp parallel for reduction(+ : unagg)
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        unagg += (agg.get_const_data()[i] == -1);
+    }
+    *num_unagg = unagg;
+}
 
 GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_AMGX_PGM_COUNT_UNAGG_KERNEL);
 
 
 template <typename IndexType>
 void renumber(std::shared_ptr<const OmpExecutor> exec, Array<IndexType> &agg,
-              IndexType *num_agg) GKO_NOT_IMPLEMENTED;
+              IndexType *num_agg)
+{
+    const auto num = agg.get_num_elems();
+    Array<IndexType> agg_map(exec, num + 1);
+    auto agg_vals = agg.get_data();
+    auto agg_map_vals = agg_map.get_data();
+    components::fill_array(exec, agg_map_vals, num + 1, zero<IndexType>());
+#pragma omp parallel for
+    for (size_type i = 0; i < num; i++) {
+        agg_map_vals[agg_vals[i]] = 1;
```

Comment:
Do I understand correctly that for each (un-remapped) coarse index `i`, `agg_vals[i] == i` holds? This kind of reminds me of a disjoint set/union find data structure. In case that's true, do you think it would make sense conceptually to make this write conditional, then? That would eliminate all duplicate writes
```suggestion
        if (agg_vals[i] == i) {
            agg_map_vals[agg_vals[i]] = 1;
        }
```

Reply:
I think so, `agg_vals[i] == i` should hold in algorithm.
It really a good idea!
agg_map_val is IndexType, so I can also use masking way to do it `(agg_val[i] == i) * 1` to eliminate the branch.

Reply:
Yes, I only proposed this solution, since it silences SonarQube, and the compiler should translate this to similar code.